File: blockchain/src/main/java/pt/um/lei/masb/blockchain/data/OtherData.java
package pt.um.lei.masb.blockchain.data;

import pt.um.lei.masb.blockchain.Sizeable;

import javax.persistence.Basic;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.validation.constraints.NotNull;
import java.io.Serializable;
import java.math.BigDecimal;
import java.util.Objects;


@Entity
public final class OtherData<T extends Serializable>
        extends GeoData
        implements Sizeable {

    @Id
    @GeneratedValue
    private long id;

    @Basic(optional = false)
    private final T data;

    protected OtherData() {
        this.data = null;
    }

    public OtherData(@NotNull T data,
                     BigDecimal lat,
                     BigDecimal lng) {
        super(lat, lng);
        this.data = data;
    }


    public T getData() {
        return data;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }
        OtherData<?> otherData = (OtherData<?>) o;
        return Objects.equals(data, otherData.data);
    }

    @Override
    public int hashCode() {

        return Objects.hash(data);
    }

    @Override
    public @NotNull String toString() {
        return "OtherData{" +
                "data=" + data +
                '}';
    }
}

File: blockchain/src/main/kotlin/pt/um/lei/masb/blockchain/data/OtherData.kt
package pt.um.lei.masb.blockchain.data

import com.orientechnologies.orient.core.record.OElement
import com.orientechnologies.orient.core.record.impl.ORecordBytes
import pt.um.lei.masb.blockchain.Hash
import pt.um.lei.masb.blockchain.persistance.NewInstanceSession
import pt.um.lei.masb.blockchain.utils.Crypter
import java.io.ByteArrayOutputStream
import java.io.ObjectOutputStream
import java.math.BigDecimal

data class OtherData(
    val data: java.io.Serializable
) : BlockChainData {
    override fun digest(c: Crypter): Hash {
        val bao = ByteArrayOutputStream(256)
        ObjectOutputStream(bao).use {
            it.writeObject(data)
        }
        return c.applyHash(
            bao.toByteArray()
        )
    }


    override val dataConstant: Int
        get() = DataDefaults.DEFAULT_UNKNOWN

    override fun store(
        session: NewInstanceSession
    ): OElement =
        session
            .newInstance("Other")
            .let { doc ->
                val byteStream =
                    ByteArrayOutputStream(
                        approximateSize.toInt()
                    )
                ObjectOutputStream(
                    byteStream
                ).use {
                    it.writeObject(data)
                    val blob = ORecordBytes(
                        byteStream.toByteArray()
                    )
                    doc.setProperty(
                        "data",
                        blob
                    )
                }
                doc
            }

    override fun calculateDiff(
        previous: SelfInterval
    ): BigDecimal =
        BigDecimal.ONE

    override fun toString(): String =
        "OtherData(data = $data)"
}
